### ISSUE TITLE
Integrate with Crowdin localization project

### ISSUE THREAD

**Issue by bvaughn** [reactjs/reactjs.org#82]
Migrated from facebook/react/issues/8063

Coordinate with docs translators via @ericnakagawa

This issue involves multiple steps:
- [x] Integrate with [Crowdin](https://crowdin.com/project/react)  to download localized content
- [ ] Generate localized URLs using said content
- [ ] Automatically sync English content changes from GitHub to Crowdin (see [this comment](https://github.com/reactjs/reactjs.org/issues/82#issuecomment-337943132) and [this repo](https://github.com/thejameskyle/crowdin-sync))
- [ ] Maintain backwards compatibility for all current (non-localized, English) links
- [ ] Add localization toggle to header
- [ ] Update the "_Edit this page_" link to point to GitHub for English pages and Crowdin for other languages.

### Integrate with Crowdin

This could be done via a custom Gatsby plug-in. We've already written [a couple](https://github.com/reactjs/reactjs.org/tree/master/plugins). (It would be nice to work with @KyleAMathews to share this plug-in back with Gatsby so that other users could benefit from it as well.)

If we run into any performance problems, @ericnakagawa has volunteered to work with Netlify and Crowdin to simplify or speed up this syncing process. Ideally this sync process will be fast enough to allow Netlify's PR previews to also be used to test localized builds of the site.

### Misc thoughts and questions
* In the case of non-localized content, Crowdin serves English as fallback.
* How can site consume list of active locales? Is there a Crowdin API?
* How does this impact urls (eg `/docs/<file>`, `/docs/<lang>/<file>`)?

**Comment by KyleAMathews** [comment]
Happy to help create a Gatsby source plugin for Crowdin! That'd be really powerful. And a lot of people would use it including sooner than later gatsbyjs.org

From a bit of research, it looks like it'd be pretty easy to pull data using their API:

* https://support.crowdin.com/api/download/
* Guide for creating a source plugin https://www.gatsbyjs.org/docs/create-source-plugin/ — this guide is a bit wordy right now. When someone wants to tackle this, I could take a pass at improving the guide + adding a mini-tutorial on creating source plugins.

I spent a bunch of time adding localization support to https://www.gatsbyjs.org/packages/gatsby-source-contentful/ as well so have some opinions on this. See the [example site for Contentful](https://using-contentful.netlify.com/) + [its source](https://github.com/gatsbyjs/gatsby/blob/e3176bd5cc843ba2e4b5d0f3216adfaf7c2f1e26/examples/using-contentful/src/pages/index.js#L84)

**Comment by bvaughn** [comment]
Fantastic 😄  There's a good chance we'll end up collaborating on this soon then.

**Comment by Daniel15** [comment]
Netlify also has a way of doing country-based redirects. For example, if someone from Italy visits the site and they don't have a language cookie set, we can automatically redirect to the Italian version. We use this on the Yarn site - You can look at how Yarn does it for inspiration

> How does this impact urls (eg /docs/<file>, /docs/<lang>/<file>)?

I'd suggest `/<lang>/docs/`, this is consistent with other Facebook projects that use localization:
https://yarnpkg.com/en/docs/getting-started
https://flow.org/en/docs/getting-started/

> If we run into any performance problems, @ericnakagawa has volunteered to work with Netlify and Crowdin to simplify or speed up this syncing process. Ideally this sync process will be fast enough to allow Netlify's PR previews to also be used to test localized builds of the site.

We have perf issues with the Yarn site, but a lot of it is because we're using Jekyll. Most of the build time is actually the Jekyll build time rather than CrowdIn. We had a prototype using Hugo that was much faster.

> From a bit of research, it looks like it'd be pretty easy to pull data using their API:

Would the React site actually need a custom API integration? It could be useful, but CrowdIn has a CLI tool (https://support.crowdin.com/cli-tool/) which we might be able to use with less effort. For Yarn we just run `crowdin-cli download -b master` to download all the localized files as part of the build (in that case, there's a bunch of Markdown files as well as some yml files for language strings). Then you just build the site like normal and everything just works, as long as the files are in the correct locations.

**Comment by bvaughn** [comment]
>> How does this impact urls (eg /docs/, /docs//)?
> I'd suggest /<lang>/ (eg. /<lang>/docs/<file>), this is consistent with other Facebook projects that use localization:

Don't have time to fully digest this response now- but wanted to mention that we should ensure whatever change we make does not break pre-existing links (like how [links broke for Jest](https://github.com/reactjs/reactjs.org/commit/174d7890ed91276c85d8e9f985fd742f61f39cca#diff-b7c667cdd69b06a052dd48c32bb585e1)).

**Comment by KyleAMathews** [comment]
>Would the React site actually need a custom API integration

It doesn't need it but it seems simpler perhaps to just query the data through graphql instead of writing it out first then pulling it in.

But yeah, definitely more straightforward to just use the cli.

Would you checkin the translated content?

**Comment by Daniel15** [comment]
> but wanted to mention that we should ensure whatever change we make does not break pre-existing links 

Jest is limited in what it can do, because it's using GitHub Pages. With Netlify, you can generate some redirects as part of the build. Just need to redirect "raw" URLs (`/docs/foo`) to the localized URL (`/en/docs/foo`), either with `/en/` hard-coded or using Netlify's language redirects that I mentioned earlier (so it'd redirect `/docs/foo` to `/<user's preferred language>/docs/foo`).

> Would you checkin the translated content?

We don't for Yarn. It's pulled in as part of the build process for the site. Since CrowdIn tracks the history for the translations, we didn't really need to also have Git track it.

Some sites do commit the translated files though. Up to you - Either way is fine.

**Comment by bvaughn** [comment]
> Would you checkin the translated content?

No

**Comment by bvaughn** [comment]
> or using Netlify's language redirects that I mentioned earlier (so it'd redirect /docs/foo to /<user's preferred language>/docs/foo).

Cool! 👍

**Comment by Daniel15** [comment]
Here's how the Yarn site builds all the redirects from non-localized URLs to localized URLs: https://github.com/yarnpkg/website/blob/master/_redirects#L48-L70. Unfortunately it's super difficult to understand because it's using the Jekyll templating syntax. Somehow @thejameskyle got it working though. You could likely make a nicer JS script to do the same thing for the React site.

The resulting redirects file looks something like this:
```
/docs/cli/add/  /en/docs/cli/add/  302  Language=en
/docs/cli/add/  /fr/docs/cli/add/  302  Language=fr
/docs/cli/add/  /pt-BR/docs/cli/add/  302  Language=pt-br
/docs/cli/add/  /zh-Hans/docs/cli/add/  302  Language=zh-cn
```
Here's the entire language redirects section: https://gist.github.com/Daniel15/d84b39e8a260ee2a7d627751d015181b

Docs: https://www.netlify.com/docs/redirects/#geoip-and-language-based-redirects

Might be able to use a regex or placeholders or something rather than having to generate all the redirects like that.

**Comment by KyleAMathews** [comment]
@bvaughn wrote a nice [redirect API for Gatsby](https://www.gatsbyjs.org/docs/bound-action-creators/#createRedirect) + added it to [gatsby-plugin-netlify](https://www.gatsbyjs.org/packages/gatsby-plugin-netlify/) so this setting up redirects would be pretty easy :-)

**Comment by smikitky** [comment]
Hi, the official doc in English is being updated frequently, but the contents in Crowdin seems to be out of sync with the newest doc. I think it's still based on the doc several months ago, and I'm seeing no doc for error boundaries ready for translation.

I'm new to Crowdin, but I want to believe there is a way to merge the updates in the original doc and enable people to translate the changed part incrementally. Is there any plan for this?

**Comment by bvaughn** [comment]
Crowdin docs show support in their REST API to [add](https://support.crowdin.com/api/add-file/), [update](https://support.crowdin.com/api/update-file/), and [delete](https://support.crowdin.com/api/delete-file/) a file- so it seems plausible that we could write a script to do this. Maybe even get CircleCI to automate it anytime there are commits to the `master` branch?

cc @ericnakagawa who has been leading the localization effort until now.

**Comment by Daniel15** [comment]
I think we already have something configured with Yarn to automate it - Crowdin is automatically updated when changes are committed to the English versions. I'm not quite sure how it works though.

**Comment by bvaughn** [comment]
Thanks Daniel. I added a note to the description of this issue to check what Yarn's doing for this.

**Comment by Daniel15** [comment]
It seems like Crowdin have some sort of Github integration now. I don't think Yarn is using this though: https://support.crowdin.com/github-integration/.

James Kyle set up the Yarn one, it might be worth reaching out to him.

**Comment by thejameskyle** [comment]
Hi hi hi, yes I use https://github.com/thejameskyle/crowdin-sync for Yarn

**Comment by clemmy** [comment]
I'm interested in helping out with this as well. 🙂 @KyleAMathews @Daniel15 Do you guys have a preference on whether we do this through Github integration or just in time (during the build)?

**Comment by Daniel15** [comment]
I don't think it really makes sense to commit translations to the repo, since CrowdIn already tracks changes to them. Either way is fine though, I don't have a strong opinion about it.

**Comment by bvaughn** [comment]
I think we have a couple of options that may be worth exploring. I have been leaning toward using the Crowdin API to pull data at build-time. If this turns out to be slow (either the API or the resulting Gatsby build time) then we can provide a mode to disable it for local dev use.

**Comment by Daniel15** [comment]
For what it's worth, on the Yarn site we only pull down localizations in
the prod build.

**Comment by bvaughn** [comment]
Agreed that we shouldn't do them for `yarn dev` but- depending on the speed- it might also be nice to disable for `yarn build`. (It's often useful to test the "production" build locally too.)

**Comment by bvaughn** [comment]
Here are some thoughts I had about creating localized links that are also backwards-compatible:

We could decorate Gatby's `createPage` action creator and require an additional language code parameter. This language code parameter could be prepended to the slug to generate a unique URL for a given locale (eg `${language}/${slug}`). If the language code specified is English, we could also automatically create a redirect (using the `createRedirect` action creator) to a no-locale slug for backwards compatibility.

Gatsby has the following plug-ins that we should check out:
* [`gatsby-plugin-i18n`](https://github.com/angeloocana/gatsby-plugin-i18n)
* [`gatsby-plugin-i18n-readnext`](https://github.com/angeloocana/gatsby-plugin-i18n-readnext)

The plug-in `gatsby-plugin-i18n` includes a `langKeyDefault` option. However I'm not sure how tested this plug-in is or how well it would work for us. However it might be worth looking at its implementation for ideas.

**Comment by bvaughn** [comment]
Realized another issue we'll need to solve while iterating on this today: How we should make links locale-aware?

The only ways I can think of are
 1. Require our translators to add a language code prefix before each link (which sucks)
 1. Write our own Remark transform to prepend the language code based on the location of the markdown file.

I think option 2 sounds more likely to work but it also sounds like a lot of effort. Will need to research and think about this more.

**Comment by Daniel15** [comment]
You shouldn't create redirects only for English, you should use Netlify's
locale-based redirects too (so if someone's browser or OS is set to Italian
and a page has an Italian translation, the local-less URL redirects to the
Italian page. This means you'll generate one redirect per page per locale,
and one extra one for English as default (if the page hasn't been
translated to a language yet).

For locale aware links, do you just mean links on the site? Just use
relative links everywhere :) I think that's what we do on the Yarn site.

Sent from my phone.


On Nov 10, 2017 8:28 AM, "Brian Vaughn" <notifications@github.com> wrote:

Realized another issue we'll need to solve while iterating on this today:
How we should make links locale-aware?

The only ways I can think of are

   1. Require our translators to add a language code prefix before each
   link (which sucks)
   2. Write our own Remark transform to prepend the language code based on
   the location of the markdown file.

I think option 2 sounds more likely to work but it also sounds like a lot
of effort. Will need to research and think about this more.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/reactjs/reactjs.org/issues/82#issuecomment-343519982>,
or mute the thread

**Comment by bvaughn** [comment]
Hey @Daniel15!

If we could use Netlify somehow to do a locale-friendly redirect for old links, that's great. I didn't know about that. That would work for people visiting Gatsby from a referrer page.

Once they're there, we'll need to make sure links within the site work (eg ones that are defined in Markdown as well as pages using the Gatsby `Link` component). Yarn's website and this one are significantly different, so what works for one doesn't necessarily apply to the other. This is particularly true with markdown links. Let's chat via Messenger about this if you'd like.

**Comment by bvaughn** [comment]
Quick update: I wrote a remark transform last night that handles one part of this:
```js
module.exports = (
  {markdownAST, markdownNode, getNode},
  {defaultLanguageCode, defaultSourceName, translationsSourceName},
) => {
  const parentNode = getNode(markdownNode.parent);
  const {relativePath, sourceInstanceName} = parentNode;

  let languageCode = null;

  switch (sourceInstanceName) {
    case defaultSourceName:
      languageCode = defaultLanguageCode;
      break;
    case translationsSourceName:
      languageCode = getLanguageCodeFromPath(relativePath);
      break;
  }

  if (languageCode !== null) {
    // Prepand language code before links with absolute URLs,
    // eg '/path/to/file.html' => '/en/path/to/file.js'
    visit(markdownAST, `link`, node => {
      if (node.url.startsWith('/')) {
        node.url = `/${languageCode}${node.url}`;
      }
    });
  }
};
```

Some quick testing locally seems to suggest this works well enough. Next up is a decorator for `Link`.

**Comment by Daniel15** [comment]
> If we could use Netlify somehow to do a locale-friendly redirect for old links, that's great. I didn't know about that.

I mentioned it in an earlier comment: https://github.com/reactjs/reactjs.org/issues/82#issuecomment-337721594

>  (I don't see it mentioned in the Netlify docs for redirects.)

It's on the page you linked to, under the "GeoIP and Language-based redirects" heading. 😛

**Comment by Daniel15** [comment]
> If we could use Netlify somehow to do a locale-friendly redirect for old links, that's great. I didn't know about that.

I mentioned it in an earlier comment: https://github.com/reactjs/reactjs.org/issues/82#issuecomment-337721594

>  (I don't see it mentioned in the Netlify docs for redirects.)

It's on the page you linked to, under the "GeoIP and Language-based redirects" heading. 😛 

Here's what the language redirects look like for the Yarn site, as an example: https://gist.github.com/Daniel15/d84b39e8a260ee2a7d627751d015181b

**Comment by bvaughn** [comment]
I guess I misunderstood the section of the docs on the page I linked to. Thanks for linking to the Yarn redirects file. The Gatsby redirects plug-in will need an update to support the additional language parameter but that should be pretty straight forward.

The `Link` component will still need the ability to detect and prepend the language code itself (during the build phase) because of how Gatsby generates and optimizes its content graph.

**Comment by bvaughn** [comment]
PR to add support to Gatsby + `gatsby-plugin-netlify` for language and country based redirects: gatsbyjs/gatsby/pull/2890

**Comment by bvaughn** [comment]
May be worth noting, explicitly, that we don't plan to support localization for pages like [the "Blog" landing page](https://github.com/reactjs/reactjs.org/blob/master/src/pages/blog/all.html.js). (There's no plans to translate blog post content.) If we _did_ decide to support that in the future, we would need to fork the page it somehow so that its GraphQL query only pulled in localized content.

**Comment by techhtml** [comment]
Hi there, I have a question about localization.
When I end translating Document into Korean, How can I see that in React site?

**Comment by bvaughn** [comment]
That's not currently possible, @techhtml. Once this issue is done, it will be.

**Comment by techhtml** [comment]
@bvaughn Thank you! i'll wait that work done.

**Comment by techhtml** [comment]
@ericnakagawa Hi Eric, I have question about crowdin localization. I translate `static-type-checking` docs into Korean. but crowdin's document are not updated that only have old one. How can I update this docs?

**Comment by clemmy** [comment]
Hey @techhtml, I believe your translations need to be reviewed and approved by an admin of the React project on Crowdin before your changes appear.

**Comment by techhtml** [comment]
@clemmy of course. Other Korean developers will help reviews my translations.

**Comment by smikitky** [comment]
Hi, I logged in to Crowdin after a while, and found the "translated" percentage of Japanese has dropped to 65%. It was 100% (i.e., translation completed) in September last year, but no one but translators themselves had a chance to read the translated docs.

So can someone set up a timeline and a guideline for this on which we can rely on? You have already gathered dozens of volunteers at Crowdin even though you have not officially called for volunteers. Has the localization project started *officially*, in the first place? Are you going to call for more collaborators, in the blog for example? I don't think the entire Crowdin project may be abandoned, but since there is no timeline nor document to guide the translators, I don't know when we can restart translating to keep up with the current doc. I don't even know what those "master" and "test-17" folders mean. (BTW, I just found the translation percentage for the "master" folder is now 1%... Does this mean almost all the effort we already made on the "old" master is going to disappear soon?)

![image](https://user-images.githubusercontent.com/7779406/35316561-adc27560-0115-11e8-9c43-7eeb451b6bf9.png)

And please also clarify which is to be translated. I'm now seeing "blog" and "conferences" folder in the Crowdin project again, which I believe is not to be translated.


cc @ericnakagawa

**Comment by bvaughn** [comment]
Sorry, I don't have any updates/estimates for this. The core team is spread very thin and I don't think anyone owns this effort at the moment 😞

**Comment by gaearon** [comment]
To give an update — if everything works out, we expect somebody to jump on finishing the integration in the following weeks. Can't promise to get it done before Hooks are stable though but it's on our radar.

**Comment by smikitky** [comment]
For the record, here are some problems I'm currently seeing as a translator.

- Translations are not Markdown-based but HTML-based. I believe all translated materials should be version-controlled.
- There are several "root" directories on the Crowdin repo, and it is not clear which to work on.
- New commits to the original doc are not automatically synched. For example there are no articles about Hooks yet in the Crowdin repository.
- Some elements in the front matter (document IDs, etc) are wrongly marked as translatable.
- Some code examples are marked as translatable and some are not.
- This may be specific to CJK languages, but italics are not always rendered properly for those languages, so we need a different CSS styles for `<em>`'s.

**Comment by Daniel15** [comment]
> Translations are not Markdown-based but HTML-based. 

For Yarn we use Markdown in CrowdIn, so this might just be an issue with how it's configured for React. Should be straightforward to fix it 😃

**Comment by tesseralis** [comment]
@smikitky We've come up with an alternate approach here: https://github.com/reactjs/reactjs.org/issues/1605

Does this resolve any of your issues working with Crowdin, and would you be interested in working with us on maintaining the Japanese fork? We should be able to use the work that you and others have done so far on Crowdin as a baseline.

**Comment by gaearon** [comment]
https://reactjs.org/blog/2019/02/23/is-react-translated-yet.html
